File: leetcode-04/solutions/cpp/0074.cpp
class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    if (matrix.empty()) return false;

    int l = 0;
    int r = matrix.size() * matrix[0].size();

    while (l < r) {
      int m = (l + r) >> 1;
      int i = m / matrix[0].size();
      int j = m % matrix[0].size();
      if (matrix[i][j] == target) return true;
      if (matrix[i][j] > target)
        r = m;
      else
        l = m + 1;
    }

    return false;
  }
};

File: leetcode-04/solutions/python3/0074.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return False

        l = 0
        r = len(matrix) * len(matrix[0])

        while l < r:
            m = (l + r) >> 1
            i = m // len(matrix[0])
            j = m % len(matrix[0])
            if matrix[i][j] == target:
                return True
            if matrix[i][j] > target:
                r = m
            else:
                l = m + 1

        return False
